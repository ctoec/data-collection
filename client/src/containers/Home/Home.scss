.Home {
  .how-to-list {
    list-style: none;
    list-style-type: none;
    counter-reset: item;
    padding: 0;
    margin: 0;
  }

  .how-to-step::before {
    content: counter(item);
    line-height: 2.5rem;
    background: #61366f;
    color: white;
    font-size: x-large;
    font-weight: bold;
    text-align: center;
    border-radius: 50%;
    position: absolute;
    height: 2.5rem;
    width: 2.5rem;
    left: 0;
    display: block;
  }

  .how-to-step {
    counter-increment: item;
    position: relative;
    padding-left: 3.5rem;

    .how-to-header {
      line-height: 2.5rem;
      font-size: x-large;
      font-weight: bold;
      margin-bottom: 0;
      padding-bottom: 0;
    }

    .how-to-body {
      margin-top: 0.25rem;
      margin-bottom: -0.5rem;
    }

    .how-to-button {
      margin-bottom: 2rem;
    }
  }
}

.three-column-layout {
  display: flex;
  flex-wrap: wrap;
  flex-direction: row;
  flex: 1;
  align-items: flex-start;
}

.three-column-card {
  margin: 0.5em;
  width: 30%;
  max-width: 30%;
}
