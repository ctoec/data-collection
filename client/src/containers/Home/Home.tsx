import React from 'react';

const Home: React.FC = () => (
	<div>This is the data collection tool</div>
);

export default Home;