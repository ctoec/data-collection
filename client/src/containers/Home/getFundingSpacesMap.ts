type enrollmentTime = {
  timeType: string;
  filled: number;
  capacity: number;
};
type enrolledAgeGroup = {
  ageGroup: string;
  includedTimes: enrollmentTime[];
};

export const getFundingSpacesDisplay = (fundingSpacesMap: any) => {
  // Structure to hold accumulated display information about each funding space
  const fundingSpacesDisplay: {
    sourceName: string;
    includedAgeGroups: enrolledAgeGroup[];
  }[] = [];

  // Server will return funding spaces separated by unique IDs, which
  // the post-submit home page will hand us, so go through them to
  // place each space in the right bucket
  fundingSpacesMap.forEach((fsCount: any) => {
    // Start with overall funding source, since it's the header
    const source = fsCount.fundingSpace.source;
    let matchingSource = fundingSpacesDisplay.find(
      (fsd) => fsd.sourceName === source
    );
    if (matchingSource === undefined) {
      matchingSource = { sourceName: source, includedAgeGroups: [] };
      fundingSpacesDisplay.push(matchingSource);
    }

    // Then the age group, since a source is made up of a list of these
    const ageGroup = fsCount.fundingSpace.ageGroup;
    let matchingGroup = matchingSource.includedAgeGroups.find(
      (ag) => ag.ageGroup === ageGroup
    );
    if (matchingGroup === undefined) {
      matchingGroup = { ageGroup, includedTimes: [] };
      matchingSource.includedAgeGroups.push(matchingGroup);
    }

    // Then go to the times within each age group, because that's the
    // last sub-list
    const timeType = fsCount.fundingSpace.time;
    let matchingTime = matchingGroup.includedTimes.find(
      (t) => t.timeType === timeType
    );
    if (matchingTime === undefined) {
      matchingTime = {
        timeType,
        filled: 0,
        capacity: fsCount.fundingSpace.capacity,
      };
      matchingGroup.includedTimes.push(matchingTime);
    }
    matchingTime.filled += 1;
  });

  return fundingSpacesDisplay;
};
