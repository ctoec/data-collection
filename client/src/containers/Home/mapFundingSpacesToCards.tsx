import React from 'react';
import { Card } from '@ctoec/component-library';

// Map each calculated funding space distribution into a card
// element that we can format for display
export const mapFundingSpacesToCards = (fundingSpacesDisplay: any) => {
  const fundingCards = (fundingSpacesDisplay || []).map((fsd: any) => (
    <div className="desktop:grid-col-4 three-column-card">
      <Card>
        <div className="padding-0">
          <h3>{fsd.sourceName}</h3>
          {fsd.includedAgeGroups.map((ag: any) => (
            <>
              <div>
                <p className="text-bold">{ag.ageGroup}</p>
              </div>
              <>
                {ag.includedTimes.map((t: any) => {
                  // Account for fundings that have negative capacities stored
                  // Also hide capacities for SR fundings, since we're less
                  // confident on those and they change a lot
                  let spaceNumbers = `${t.filled}`;
                  if (
                    t.capacity !== -1 &&
                    !fsd.sourceName.includes('School Readiness')
                  )
                    spaceNumbers += `/${t.capacity} `;
                  else spaceNumbers += ` `;
                  return (
                    <p className="text-base-darker padding-0 margin-0">
                      <b>{spaceNumbers}</b>
                      {` ${t.timeType}`}
                    </p>
                  );
                })}
              </>
            </>
          ))}
        </div>
      </Card>
    </div>
  ));
  return fundingCards;
};
