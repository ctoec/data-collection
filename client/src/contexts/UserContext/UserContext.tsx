import React, {
  Dispatch,
  SetStateAction,
  useContext,
  useEffect,
  useState,
} from 'react';
import AuthenticationContext from '../AuthenticationContext/AuthenticationContext';
import { User } from '../../shared/models';
import { apiGet, apiPost } from '../../utils/api';
import { Moment as MomentInterface } from 'moment';
import * as moment from 'moment';

export type UserContextType = {
  user: User | null;
  loading: boolean;
  setConfidentialityAgreedDate: Dispatch<SetStateAction<Date>>;
  confidentialityAgreedDate: MomentInterface | null;
};

const UserContext = React.createContext<UserContextType>({
  user: null,
  loading: true,
  confidentialityAgreedDate: null,
  setConfidentialityAgreedDate: () => {},
});

const { Provider, Consumer } = UserContext;

export type UserProviderPropsType = {};

/**
 * Context Provider for injecting User object into React hierarchy
 *
 * @param props Props with user
 */
const UserProvider: React.FC<UserProviderPropsType> = ({ children }) => {
  const { accessToken, loading } = useContext(AuthenticationContext);
  const [user, setUser] = useState<User | null>(null);
  const [userLoading, setUserLoading] = useState(loading);
  const [refetchUser, setRefetchUser] = useState(false);

  useEffect(() => {
    if (accessToken) {
      setUserLoading(true);
      apiGet('users/current', accessToken)
        .then((data) => {
          setUser(data);
        })
        .catch((err) => {
          // This doesn't need to throw an error because the lack of user is taken to mean that this failed
          console.error(err);
        })
        .finally(() => {
          setUserLoading(false);
          setRefetchUser(false);
        });
    } else {
      setUserLoading(loading);
    }
  }, [accessToken, loading, refetchUser]);

  function setConfidentialityAgreedDate() {
    apiPost(
      `users/current`,
      { ...user, confidentialityAgreedDate: moment.utc() },
      { accessToken, jsonParse: false }
    )
      .then(() => setRefetchUser(true))
      .catch((err) => {
        console.error(err);
        throw new Error('Error processing confidentiality agreement');
      });
  }

  return (
    <Provider
      value={{
        loading: userLoading,
        user,
        confidentialityAgreedDate: user?.confidentialityAgreedDate || null,
        setConfidentialityAgreedDate,
      }}
    >
      {children}
    </Provider>
  );
};

export { UserProvider };
export { Consumer as UserConsumer };
export default UserContext;
