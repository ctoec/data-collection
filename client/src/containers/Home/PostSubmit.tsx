import React, { useContext, useEffect, useState } from 'react';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import { apiGet } from '../../utils/api';
import pluralize from 'pluralize';
import { AddRecordButton } from '../../components/AddRecordButton';
import { Link } from 'react-router-dom';
import {
  Alert,
  ArrowRight,
  Card,
  InlineIcon,
  TextWithIcon,
  Divider,
  Button,
} from '@ctoec/component-library';
import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
import UserContext from '../../contexts/UserContext/UserContext';
import { mapFundingSpacesToCards } from './utils/mapFundingSpacesToCards';
import { NestedFundingSpaces } from '../../shared/payloads/NestedFundingSpaces';

export const PostSubmitHome: React.FC = () => {
  const { user } = useContext(UserContext);
  const { accessToken } = useContext(AuthenticationContext);
  const h1Ref = getH1RefForTitle();
  const orgAccess = user?.accessType === 'organization';
  const userOrgs = user?.organizations || [];
  const showFundings = orgAccess && userOrgs.length == 1;

  // Count how many children are in the roster so we can format the display
  const [userRosterCount, setUserRosterCount] = useState(undefined);
  const [
    fundingSpacesDisplay,
    setFundingSpacesDisplay,
  ] = useState<NestedFundingSpaces>();
  const [siteCountDisplay, setSiteCountDisplay] = useState();
  useEffect(() => {
    apiGet(`children/metadata?showFundings=${showFundings}`, accessToken)
      .then(({ count, fundingSpacesMap, siteCountMap }) => {
        setUserRosterCount(count);
        // Get the site count data structure, too, so we can divide up
        // user-accessible sites into individual cards
        setSiteCountDisplay(siteCountMap);

        // Also determine the funding spaces map for the organization, if
        // the user has the permissions that enable this
        if (fundingSpacesMap) {
          setFundingSpacesDisplay(fundingSpacesMap);
        }
      })
      .catch((err) => {
        throw new Error(err);
      });
  }, [accessToken]);

  return (
    <div className="grid-container margin-top-4">
      <h1 ref={h1Ref} className="margin-top-0">
        Hello {user?.firstName}!
      </h1>
      <div className="margin-top-3 margin-bottom-1">
        <Divider />
      </div>
      {userOrgs.map((org) => (
        <div className="grid-row grid-gap">
          <div className="tablet:grid-col-10">
            <h2>{org.providerName}</h2>
            <p className="font-body-lg text-base-darker">
              {`${userRosterCount} active ${pluralize(
                'enrollment',
                userRosterCount,
                false
              )} across ${pluralize('site', (user?.sites || []).length, true)}`}
            </p>
            <AddRecordButton
              id="home-page-add-enrollment"
              appearance="default"
            />
          </div>
          <div className="margin-top-4 tablet:grid-col-2">
            <Button
              appearance="unstyled"
              href="/roster"
              text={
                <TextWithIcon
                  text="View roster"
                  iconSide="right"
                  Icon={ArrowRight}
                  direction="right"
                  className="text-underline"
                />
              }
            />
          </div>
        </div>
      ))}
      <div className="margin-top-4 margin-bottom-4">
        <Divider />
      </div>
      <h2>Updates and tasks</h2>
      <Alert
        type="info"
        text={
          <span>
            Are your sites and/or funding spaces incorrect? Reach out to the ECE
            Reporter team through&nbsp;
            <Link to="/revision">this form.</Link>
          </span>
        }
      />
      <div className="grid-row margin-top-2">
        <Card>
          <div className="grid-container padding-0">
            <span className="grid-row font-body-lg text-bold">
              <div className="text-center">
                <InlineIcon icon="complete" />
              </div>
              Thank you for completing your July to December 2020 data
              collection!
            </span>
            <p className="grid-row">
              OEC appreciates all that you do to improve the lives of children
              and families.
            </p>
            <h3 className="grid-row">What's next?</h3>
            <p className="grid-row margin-bottom-1em">
              Keep your roster updated with new enrollments and withdrawals, and
              with age group, site, and funding changes.
            </p>
            <p className="grid-row">
              We're working on a feature to allow updating from excel files.
              Until then, you can make any needed changes directly in your ECE
              Reporter roster.
            </p>
          </div>
        </Card>
      </div>
      <div className="margin-top-4 margin-bottom-4">
        <Divider />
      </div>
      {showFundings && fundingSpacesDisplay && (
        <>
          <h2>Funding spaces</h2>
          <div className="three-column-layout">
            {mapFundingSpacesToCards(fundingSpacesDisplay)}
          </div>
          <div className="margin-top-4 margin-bottom-4">
            <Divider />
          </div>
        </>
      )}
      <h2>Sites</h2>
      <div className="three-column-layout">
        {(siteCountDisplay || []).map((s: any) => (
          <div className="desktop:grid-col-4 three-column-card">
            <Card>
              <div className="padding-0">
                <h3>{s.siteName}</h3>
                <p className="text-base-darker">
                  {pluralize('enrollment', s.count, true)}
                </p>
                <Link to={`/roster?organization=${s.orgId}&site=${s.siteId}`}>
                  <TextWithIcon
                    text="View site roster"
                    iconSide="right"
                    Icon={ArrowRight}
                    direction="right"
                  />
                </Link>
              </div>
            </Card>
          </div>
        ))}
      </div>
    </div>
  );
};
