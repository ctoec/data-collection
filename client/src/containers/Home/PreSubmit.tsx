import { Alert } from '@ctoec/component-library';
import Divider from '@material-ui/core/Divider';
import React, { useContext, useEffect, useState } from 'react';
import { Link } from 'react-router-dom';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import UserContext from '../../contexts/UserContext/UserContext';
import { useAlerts } from '../../hooks/useAlerts';
import { ReactComponent as Image } from '../../images/PersonWithSpreadsheet.svg';
import { apiGet } from '../../utils/api';
import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
import { HowToUseStep } from './HowToUseStep';
import { mapFundingSpacesToCards } from './utils/mapFundingSpacesToCards';

export const PreSubmitHome: React.FC = () => {
  const { user } = useContext(UserContext);
  const { accessToken } = useContext(AuthenticationContext);
  const h1Ref = getH1RefForTitle();
  const orgAccess = user?.accessType === 'organization';
  const userOrgs = user?.organizations || [];
  const showFundingsAndSites = orgAccess && userOrgs.length == 1;

  // We might have a success alert pushed from the revision request
  // form, so check whether we do (but filter so that we only show
  // one if a user submits multiple forms)
  const [alertFromRevision, setAlertFromRevision] = useState<boolean>();
  const { setAlerts, alertElements } = useAlerts();
  // let alertToShow;
  useEffect(() => {
    setAlerts((_alerts) => [
      _alerts.find((a) => a?.heading === 'Request received!'),
    ]);
    // alertToShow = alerts.find((a) => a?.heading === "Request received!");
    // console.log(alertToShow);
    // if (alertToShow) setAlerts([alertToShow]);
  }, []);

  console.log(alertElements);
  const [fundingSpacesDisplay, setFundingSpacesDisplay] = useState();

  useEffect(() => {
    // Determine the funding spaces map for the organization, if
    // the user has the permissions that enable this
    if (showFundingsAndSites) {
      apiGet('children?fundingMap=true', accessToken)
        .then((res) => {
          setFundingSpacesDisplay(res.fundingSpacesMap);
        })
        .catch((err) => {
          throw new Error(err);
        });
    }
  }, [accessToken, showFundingsAndSites]);

  let fundingCards = mapFundingSpacesToCards(fundingSpacesDisplay);
  const fundingSection = (
    <>
      <h3 className="pre-submit-h3">Funding spaces</h3>
      <div className="three-column-layout">{fundingCards}</div>
      <div className="margin-top-4 margin-bottom-4">
        <Divider />
      </div>
    </>
  );

  const siteSection = (
    <>
      <h3 className="pre-submit-h3">Sites</h3>
      <ul>
        {(user?.sites || []).map((site) => (
          <li key={site.siteName} className="line-height-body-4">
            {site.siteName}
          </li>
        ))}
      </ul>
    </>
  );

  return (
    <div className="Home grid-container margin-top-4">
      {alertElements.length > 0 && alertElements}
      <div className="grid-row grid-gap">
        <div className="tablet:grid-col-8 margin-top-4">
          <h1 ref={h1Ref} className="margin-top-0">
            Hello {user?.firstName}!
          </h1>
          <p>
            ECE Reporter will allow you to share your state-funded enrollment
            data with the Office of Early Childhood (OEC). Make sure the site
            names and funding spaces below are accurate before you start
            entering data.
          </p>
          <h2 className="pre-submit-h2">
            Your managed sites and funding spaces
          </h2>
        </div>
        <div className="tablet:grid-col-4 margin-top-4 img" role="presentation">
          <Image />
        </div>
      </div>
      <Alert
        type="info"
        text={
          <span>
            Are your sites and/or funding spaces incorrect? Reach out to the ECE
            Reporter team through &nbsp;
            <Link to="/revision">this form.</Link>
          </span>
        }
      />
      {showFundingsAndSites && siteSection}
      {showFundingsAndSites && fundingSection}
      <h2 className="pre-submit-h2">How to use ECE Reporter</h2>
      <p>
        {
          "These instructions guide you through the file upload process to add many records at once. If you'd like to add records manually, "
        }
        <Link to="/roster/"> you can add them on your roster</Link>.
      </p>
      <ol className="how-to-list">
        <li key="review-requirements" className="how-to-step">
          <HowToUseStep
            stepHeader="Review the data requirements"
            paragraphText="Learn more about the data elements ECE Reporter collects for each enrollment."
            buttonText="See required data"
            linkTo="/data-requirements/"
          />
        </li>
        <li key="download-template" className="how-to-step">
          {
            <HowToUseStep
              stepHeader="Download the data template"
              paragraphText="You'll need to enter data in this file to upload in ECE Reporter."
              buttonText="Download the template"
              linkTo="/template/"
            />
          }
        </li>
        <li key="prepare-data" className="how-to-step">
          {
            <HowToUseStep
              stepHeader="Prepare your data"
              paragraphText="Get instructions on how to enter data into OEC's template."
              buttonText="See the how-to guide"
              linkTo="https://help.ece-reporter.ctoec.org/prepare-your-data/"
            />
          }
        </li>
        <li key="upload" className="how-to-step">
          {
            <HowToUseStep
              stepHeader="Upload your data"
              paragraphText="Once your template is filled out, you'll upload the data to ECE Reporter."
              buttonText="Go to file upload"
              linkTo="/upload/"
            />
          }
        </li>
      </ol>
    </div>
  );
};
