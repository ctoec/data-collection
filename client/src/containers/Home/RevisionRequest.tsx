import { Button, Checkbox, TextInput } from '@ctoec/component-library';
import Divider from '@material-ui/core/Divider';
import React, { useContext, useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import { BackButton } from '../../components/BackButton';
import AuthenticationContext from '../../contexts/AuthenticationContext/AuthenticationContext';
import UserContext from '../../contexts/UserContext/UserContext';
import { FUNDING_SOURCE_TIMES } from '../../shared/constants';
import { AgeGroup, FundingSpace, Site } from '../../shared/models';
import { AddSiteRequest } from '../../shared/models/db/AddSiteRequest';
import { ChangeFundingSpaceRequest } from '../../shared/models/db/ChangeFundingSpaceRequest';
import { UpdateSiteRequest } from '../../shared/models/db/UpdateSiteRequest';
import { apiGet, apiPost } from '../../utils/api';
import { getH1RefForTitle } from '../../utils/getH1RefForTitle';
import { getAddNewSiteForm } from './utils/getAddNewSiteForm';
import { getFundingSpaceCheckboxes } from './utils/getFundingSpaceCheckboxes';

/**
 * Form to allow a user to request changes to their accessible sites
 * and/or funding spaces. FormField elements aren't used here because
 * we're not modifying an existing DB entity, only accumulating
 * information that we'll shove into a new row in the revisions
 * table.
 */
export const RevisionRequest: React.FC = () => {
  const { user } = useContext(UserContext);
  const userOrgs = user?.organizations || [];
  const { accessToken } = useContext(AuthenticationContext);
  const history = useHistory();
  const h1Ref = getH1RefForTitle();

  const userSites = user?.sites || [];
  const [updateRequests, setUpdateRequests] = useState<UpdateSiteRequest[]>([]);
  useEffect(() => {
    if (user) {
      // Store all the sites the user works with in local state
      // so that we can easily modify them and send a single
      // change map to the backend
      const mappedSites = userSites.map(
        (s) =>
          ({
            siteId: s.id,
            newName: '',
            remove: false,
          } as UpdateSiteRequest)
      );
      setUpdateRequests(mappedSites);
    }
  }, [user]);

  const [addRequests, setAddRequests] = useState<AddSiteRequest[]>([]);

  // Start by determining all potential funding spaces so we know what
  // we need to precheck
  const [userFundingSpaces, setUserFundingSpaces] = useState<
    ChangeFundingSpaceRequest[]
  >([]);
  useEffect(() => {
    const fsOptions: ChangeFundingSpaceRequest[] = [];
    Object.values(AgeGroup).map((ag) => {
      FUNDING_SOURCE_TIMES.forEach((fst) => {
        if (!fst.ageGroupLimitations || fst.ageGroupLimitations.includes(ag)) {
          fst.fundingTimes.forEach((time) => {
            const stringRep = ag + ' - ' + fst.displayName + ' - ' + time.value;
            fsOptions.push({
              fundingSpace: stringRep,
              shouldHave: false,
            } as ChangeFundingSpaceRequest);
          });
        }
      });
    });
    setUserFundingSpaces(fsOptions);
  }, []);

  // Now fetch all the funding spaces that the user and their org
  // have access to, and when we generate the funding space
  // checkboxes on the page, mark all these as true
  const [fundingCheckboxes, setFundingCheckboxes] = useState<JSX.Element[]>([]);
  useEffect(() => {
    if (user && accessToken && userFundingSpaces.length !== 0) {
      apiGet('funding-spaces', accessToken)
        .then((res: FundingSpace[]) => {
          res.forEach((fs) => {
            const stringRep =
              fs.ageGroup +
              ' - ' +
              fs.source.split('-')[1].trim() +
              ' - ' +
              fs.time;
            const match = userFundingSpaces.find(
              (fs) => fs.fundingSpace === stringRep
            );
            if (match) {
              setUserFundingSpaces((o) => {
                match.shouldHave = true;
                return o;
              });
            }
          });
          setFundingCheckboxes(
            getFundingSpaceCheckboxes(userFundingSpaces, setUserFundingSpaces)
          );
        })
        .catch((err) => {
          throw new Error(err);
        });
    }
  }, [user, accessToken, userFundingSpaces.length]);

  const updateSite = (
    s: Site,
    opts?: {
      updatedName?: string;
      removeSite?: boolean;
    }
  ) => {
    const { updatedName, removeSite } = opts || {};
    setUpdateRequests((o) => {
      const site = updateRequests.find((_site) => _site.siteId === s.id);
      if (site && updatedName) site.newName = updatedName;
      if (site && removeSite) site.remove = removeSite;
      return o;
    });
  };

  // Create the nice tabular format of sites the user has
  // access to, along with the text input fields in which
  // they can change those site names, plus the checkboxes
  // to request no longer needing that site
  const siteSection = (
    <>
      <div className="grid-row grid-gap margin-bottom-1">
        <div className="tablet:grid-col-4 text-bold">Site name</div>
        <div className="tablet:grid-col-3 text-bold">Request updated name</div>
        <div className="tablet:grid-col-3 text-bold">
          Request removal from your account
        </div>
      </div>
      {(user?.sites || []).map((s) => (
        <>
          <div className="grid-row grid-gap margin-bottom-1">
            <div className="tablet:grid-col-4">
              <p>{s.siteName}</p>
              <p className="usa-hint">
                <b>License #: </b> {` ${s.licenseNumber || 'Exempt'} `}{' '}
                <b>NAEYC ID: </b> {` ${s.naeycId || 'None'} `}{' '}
                <b>Registry ID: </b> {` ${s.registryId || 'None'} `}
              </p>
            </div>
            <div className="tablet:grid-col-3 display-flex flex-align-center">
              <TextInput
                label=""
                id={`${s.siteName}-rename`}
                type="input"
                onChange={(e) => {
                  const change = e.target.value;
                  updateSite(s, { updatedName: change });
                  return change;
                }}
              />
            </div>
            <div className="tablet:grid-col-3 display-flex flex-align-center">
              <Checkbox
                id={`${s.siteName}-remove-checkbox`}
                text=""
                onChange={(e) => {
                  const checked = e.target.checked;
                  updateSite(s, { removeSite: checked });
                  return checked;
                }}
              />
            </div>
          </div>
          <div className="tablet:grid-col-10 margin-top-1 margin-bottom-1">
            <Divider />
          </div>
        </>
      ))}
    </>
  );

  // Add another set of new site fields to the revision form
  const addNewSite = () => {
    const newSite = {
      siteName: '',
      licenseId: '',
      naeycId: '',
      registryId: '',
    } as AddSiteRequest;
    setAddRequests((o) => {
      return [...o, newSite];
    });
  };

  const addSiteSection = addRequests.map((addReq, idx) =>
    getAddNewSiteForm(addReq, idx, setAddRequests)
  );

  // Send the accumulated changes to be written to the DB
  const submitRequest = () => {
    const revisionRequest = {
      updateSiteRequests: updateRequests.filter(
        (r) => r.newName !== '' || r.remove
      ),
      addSiteRequests: addRequests.filter(
        (r) => r.siteName && r.siteName !== ''
      ),
      fundingSpaceRequests: userFundingSpaces,
    };
    apiPost(`revision-request/${userOrgs[0].id}`, revisionRequest, {
      accessToken,
      jsonParse: false,
    }).catch((err) => {
      throw new Error(err);
    });
    history.push('/home', {
      alerts: [
        {
          type: 'success',
          heading: 'Request received!',
          text:
            'The ECE Reporter team will send any questions and status updates to the email associated with your ECE Reporter account.',
        },
      ],
    });
  };

  return (
    <div className="grid-container margin-top-4">
      <BackButton />
      <h1 ref={h1Ref}>Request site and funding changes</h1>
      <p>
        The ECE Reporter team will review any requested updates made in this
        form. Any updates or follow-up questions will be sent to the email
        associated with your ECE Reporter account.
      </p>
      <h2>Managed sites</h2>
      {siteSection}
      <div className="tablet:grid-col-6">
        <h2>Request a new site</h2>
        <p className="usa-hint">
          Fill out this section if your program manages a site not included in
          your ECE Reporter account.
        </p>
        {addSiteSection}
        <Button appearance="default" text="Add New Site" onClick={addNewSite} />
      </div>
      <div className="tablet:grid-col-6 margin-top-4 margin-bottom-4">
        <Divider />
      </div>
      <div className="tablet:grid-col-6">
        <h2>Funding space types</h2>
        <p className="usa-hint">
          Request corrections to the funding types received by your program in
          the section below.
        </p>
        {fundingCheckboxes}
      </div>
      <div className="grid-row grid-gap margin-top-2 margin-bottom-2">
        <Button appearance="outline" text="Cancel" href="/home" />
        <Button
          appearance="default"
          text="Send request"
          onClick={submitRequest}
        />
      </div>
    </div>
  );
};
