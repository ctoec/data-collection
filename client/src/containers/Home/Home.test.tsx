import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import Home from './Home';
import { snapshotTestHelper, accessibilityTestHelper } from '../../testHelpers';
import UserContext from '../../contexts/UserContext/UserContext';

const routerWrapped = (
  <BrowserRouter>
    <UserContext.Provider value={{ loading: false, user: null }}>
      <Home />
    </UserContext.Provider>
  </BrowserRouter>
);
describe('Home', () => {
  snapshotTestHelper(routerWrapped);
  accessibilityTestHelper(routerWrapped);
});
