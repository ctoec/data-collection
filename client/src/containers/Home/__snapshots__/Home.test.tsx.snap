// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <div
      class="usa-hero"
    >
      <div>
        <h1
          class="margin-bottom-2"
        >
          Upload your enrollment data
        </h1>
        <a
          class="usa-button usa-button--base btn--inverse"
          href="/login"
        >
          Sign in
        </a>
      </div>
    </div>
    <div
      class="grid-container margin-top-4"
    >
      <div
        class="grid-row"
      >
        <div
          class="tablet:grid-col-auto margin-right-4 padding-4"
        >
          <img
            alt=""
            class="hero-bubble"
            src="homeCareerBubble.png"
          />
        </div>
        <div
          class="tablet:grid-col-fill"
        >
          <h2
            class="text-primary text-light margin-y-3"
          >
            We support affordable child care in Connecticut
          </h2>
          <p
            class="line-height-sans-5"
          >
            Publicly-funded early care and education programs use ECE Reporter to share data with the Connecticut Office of Early Childhood.
          </p>
          <p>
            The Office of Early Childhood uses this data to pay programs and help families access quality care.
          </p>
          <div
            class="margin-bottom-4"
          >
            <p
              class="text-bold"
            >
              Learn more
            </p>
            <a
              class="usa-button usa-button--unstyled text-bold margin-bottom-3 display-block"
              href="https://ctoec.org"
            >
              <span
                class="oec-text-with-icon text-underline"
              >
                Visit OEC's website
                <svg
                  class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-right oec-text-with-icon__icon--side-right"
                >
                  arrow-right.svg
                </svg>
              </span>
            </a>
            <a
              class="usa-button usa-button--unstyled text-bold margin-bottom-3 display-block"
              href="/privacy-policy"
            >
              <span
                class="oec-text-with-icon text-underline"
              >
                See the privacy policy
                <svg
                  class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-right oec-text-with-icon__icon--side-right"
                >
                  arrow-right.svg
                </svg>
              </span>
            </a>
            <a
              class="usa-button usa-button--unstyled text-bold margin-bottom-3 display-block"
              href="/data-requirements"
            >
              <span
                class="oec-text-with-icon text-underline"
              >
                See the data requirements
                <svg
                  class="oec-text-with-icon__icon oec-text-with-icon__icon--direction-right oec-text-with-icon__icon--side-right"
                >
                  arrow-right.svg
                </svg>
              </span>
            </a>
          </div>
          <div>
            <p
              class="text-bold"
            >
              Download the data collection template
            </p>
            <button
              class="usa-button usa-button--unstyled text-bold margin-bottom-3 display-block"
              type="button"
            >
              <span
                class="oec-text-with-icon text-underline"
              >
                Download Excel template
              </span>
            </button>
            <button
              class="usa-button usa-button--unstyled text-bold margin-bottom-3 display-block"
              type="button"
            >
              <span
                class="oec-text-with-icon text-underline"
              >
                Download CSV template
              </span>
            </button>
          </div>
        </div>
      </div>
    </div>
    <footer
      class="bg-base-lightest padding-y-6 footer"
    >
      <div
        class="grid-row flex-justify flex-align-center"
      >
        <div
          class="grid-col-10"
        >
          <h2
            class="text-primary text-light margin-y-3"
          >
            Have feedback about OEC's data collection process?
          </h2>
          <p>
            Feedback from providers like you helps make this tool even better.
          </p>
          <p>
            Send us your feedback at 
            <a
              class="usa-button usa-button--unstyled"
              href="mailto:oec-data-pilot@skylight.digital?subject=Data%20collection%20bug"
              rel="noopener noreferrer"
              target="_blank"
            >
              oec-data-pilot@skylight.digital
            </a>
          </p>
        </div>
      </div>
      <div
        class="grid-col-2"
      />
    </footer>
  </div>
</DocumentFragment>
`;
